lib/src/models/document_reference.dart

```diff
@@ -0,0 +1,73 @@
+import 'package:firedart/src/repository/firestore_gateway.dart';
+import 'package:grpc/grpc.dart';
+
+import 'collection_reference.dart';
+import 'document.dart';
+import 'reference.dart';
+
+/// A [DocumentReference] refers to a document location in a Firestore database
+/// and can be used to write, read, or listen to the location.
+///
+/// The document at the referenced location may or may not exist.
+/// A [DocumentReference] can also be used to create a [CollectionReference]
+/// to a subcollection.
+class DocumentReference extends Reference {
+  /// Constructs a [DocumentReference] using [FirestoreGateway] and path.
+  ///
+  /// Throws [Exception] if path contains even amount of '/'.
+  DocumentReference(FirestoreGateway gateway, String path)
+      : super(gateway, path) {
+    if (fullPath.split('/').length % 2 == 0) {
+      throw Exception('Path is not a document: $path');
+    }
+  }
+
+  /// Returns the reference of a collection contained inside of this
+  /// document.
+  CollectionReference collection(String id) {
+    return CollectionReference(gateway, '$path/$id');
+  }
+
+  /// Reads the document referenced by this [DocumentReference].
+  ///
+  /// If no document exists, the read throw [GrpcError].
+  Future<Document> get() => gateway.getDocument(fullPath);
+
+  @Deprecated('Use the stream getter instead')
+  Stream<Document> subscribe() => stream;
+
+  /// Notifies of document by this reference.
+  Stream<Document> get stream => gateway.streamDocument(fullPath);
+
+  /// Checks if a document by this reference exists.
+  Future<bool> get exists async {
+    try {
+      await get();
+      return true;
+    } on GrpcError catch (e) {
+      if (e.code == StatusCode.notFound) {
+        return false;
+      } else {
+        rethrow;
+      }
+    }
+  }
+
+  /// Creates a document by this reference if it doesn't exist,
+  /// otherwise throw an exception.
+  Future<Document> create(Map<String, dynamic> map) => gateway.createDocument(
+      fullPath.substring(0, fullPath.lastIndexOf('/')), id, encodeMap(map));
```

Comment:
```suggestion
  /// Creates a document by this reference if it doesn't exist,
  /// otherwise throws an exception.
  Future<Document> create(Map<String, dynamic> map) => gateway.createDocument(
      fullPath.substring(0, fullPath.lastIndexOf('/')), id, encodeMap(map));
```
I'm not sure we should document throwing an exception here since there is no throwing in this method. I suppose that `gateway.createDocument` throws but this shouldn't be noticed for this method. 
@illia-romanenko, what do you think? There are several methods that contain the same confusing docs as well.